From: Nikita Popov <nikita.ppv@gmail.com>
Date: Tue, 20 Oct 2020 15:32:01 +0200
Subject: Fixed bug #80256

Remove the transfer_encoding stream filter immediately when we
destroy the old stream, to make sure it doesn't get attached to
the new stream.

Origin: https://github.com/php/php-src/commit/1c157d3fa2e102ccf375ec4c1ddca8770208dae7
Bug: https://bugs.php.net/bug.php?id=80256
---
 ext/standard/http_fopen_wrapper.c     |  9 +++++----
 ext/standard/tests/http/bug80256.phpt | 29 +++++++++++++++++++++++++++++
 2 files changed, 34 insertions(+), 4 deletions(-)
 create mode 100644 ext/standard/tests/http/bug80256.phpt

Index: php7.4-7.4.3/ext/standard/http_fopen_wrapper.c
===================================================================
--- php7.4-7.4.3.orig/ext/standard/http_fopen_wrapper.c
+++ php7.4-7.4.3/ext/standard/http_fopen_wrapper.c
@@ -853,6 +853,11 @@ finish:
 		php_stream_close(stream);
 		stream = NULL;
 
+		if (transfer_encoding) {
+			php_stream_filter_free(transfer_encoding);
+			transfer_encoding = NULL;
+		}
+
 		if (location[0] != '\0') {
 
 			char new_path[HTTP_HEADER_BLOCK_SIZE];
@@ -969,10 +974,6 @@ out:
 		if (transfer_encoding) {
 			php_stream_filter_append(&stream->readfilters, transfer_encoding);
 		}
-	} else {
-		if (transfer_encoding) {
-			php_stream_filter_free(transfer_encoding);
-		}
 	}
 
 	return stream;
Index: php7.4-7.4.3/ext/standard/tests/http/bug80256.phpt
===================================================================
--- /dev/null
+++ php7.4-7.4.3/ext/standard/tests/http/bug80256.phpt
@@ -0,0 +1,29 @@
+--TEST--
+Bug #80256: file_get_contents strip first line with chunked encoding redirect
+--SKIPIF--
+<?php require 'server-8.0.inc'; http_server_skipif(); ?>
+--INI--
+allow_url_fopen=1
+--FILE--
+<?php
+require 'server-8.0.inc';
+
+$responses = array(
+    "data://text/plain,HTTP/1.1 302 Moved Temporarily\r\n"
+    . "Location: /try-again\r\n"
+    . "Transfer-Encoding: chunked\r\n\r\n"
+    . "0\r\n\r\n",
+    "data://text/plain,HTTP/1.1 200 Ok\r\n"
+    . "Transfer-Encoding: chunked\r\n\r\n"
+    . "4\r\n1234\r\n0\r\n\r\n",
+);
+
+['pid' => $pid, 'uri' => $uri] = http_server($responses, $output);
+
+var_dump(file_get_contents($uri));
+
+http_server_kill($pid);
+
+?>
+--EXPECT--
+string(4) "1234"
